src/filters.rs

```diff
@@ -665,11 +682,46 @@ fn last_chain(rest: Filter, filter: Filter) -> (Filter, Filter) {
     }
 }
 
+/*
+ * Atempt to create an alternative representaion of a filter AST that is most
+ * suitable for fast evaluation and cache reuse.
+ */
 fn optimize(filter: Filter) -> Filter {
+    let mut filter = filter;
+    loop {
+        let pretty = simplify(filter);
+        let optimized = optimization_iterate(filter);
+        filter = simplify(optimized);
+
+        if filter == pretty {
+            return optimization_iterate(filter);
+        }
+    }
+}
+
+/*
+ * Apply optimization steps to a filter until it converges (no rules apply anymore)
+ */
+fn optimization_iterate(filter: Filter) -> Filter {
+    let mut filter = filter;
+    loop {
+        let optimized = optimization_step(filter);
+        if filter == optimized {
```

Comment:
hopefully it always converges ;D

Reply:
Oh I had some bugs of optimization steps fighting each other resulting in infinite loop.
Hopefully got them all.
This might be a good thing to fuzz..